src/ore/src/metrics.rs

```diff
@@ -403,5 +410,369 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
+pub trait MetricsFutureExt<F> {
+    /// Records the amount of time it takes a [`Future`] to complete according to "the clock on
+    /// the wall".
+    ///
+    /// More specifically, it records the instant at which the `Future` was first polled, and the
+    /// instant at which the `Future` completes. Then reports the duration between those two
+    /// instances to the provided histogram.
+    fn record_wall_time(
+        self,
+        histogram: prometheus::HistogramVec,
+        unit: TimeUnit,
+    ) -> WallTimeFuture<F>;
+
+    /// Records the total amount of time for which a [`Future`] was executing.
+    ///
+    /// More specifically, every time the `Future` is polled it records how long that individual
+    /// call took, and maintains a running sum until the `Future` completes. Then we report that
+    /// duration to the provided histogram.
+    fn record_exec_time(
+        self,
+        histogram: prometheus::HistogramVec,
+        unit: TimeUnit,
+    ) -> ExecTimeFuture<F>;
+}
+
+impl<F: Future> MetricsFutureExt<F> for F {
+    fn record_wall_time(
+        self,
+        histogram: prometheus::HistogramVec,
+        unit: TimeUnit,
+    ) -> WallTimeFuture<F> {
+        WallTimeFuture {
+            histogram,
+            unit,
+            fut: self,
+            labels: None,
+            start: None,
+            min_threshold: None,
+        }
+    }
+
+    fn record_exec_time(
+        self,
+        histogram: prometheus::HistogramVec,
+        unit: TimeUnit,
+    ) -> ExecTimeFuture<F> {
+        ExecTimeFuture {
+            fut: self,
+            histogram,
+            unit,
+            running_duration: Duration::from_millis(0),
+            labels: None,
+            min_threshold: None,
+        }
+    }
+}
+
+/// Future returned by [`MetricsFutureExt::record_wall_time`].
+#[derive(Debug)]
+#[pin_project]
+pub struct WallTimeFuture<F> {
+    /// The inner [`Future`] that we're recording the wall time for.
+    #[pin]
+    fut: F,
+    /// Prometheus histogram that we'll report to.
+    histogram: prometheus::HistogramVec,
+    /// Unit of time that we should use to report the duration.
+    unit: TimeUnit,
+    /// [`Instant`] at which the [`Future`] was first polled.
+    start: Option<Instant>,
+    /// Labels to include when recording the duration of the [`Future`].
+    labels: Option<&'static [&'static str]>,
+    /// [`Future`] durations that are less than this threshold will not get reported.
+    min_threshold: Option<Duration>,
+}
+
+impl<F> WallTimeFuture<F> {
+    /// Records the duration of this [`Future`] with the provided labels.
+    pub fn with_label_values(mut self, labels: &'static [&'static str]) -> Self {
+        self.labels = Some(labels);
+        self
+    }
+
+    /// Sets a minimum threshold. Any [`Future`]s whose duration is less than the provided
+    /// threshold will not get reported.
+    pub fn with_min_threshold(mut self, threshold: Duration) -> Self {
```

Comment:
This feels overly specific to me... it seems like a relatively rare case to want to pre-filter which durations get recorded. For instance with a histogram, wouldn't all of the short-lived events get bucketed into some small bucket, so later analysis can exclude that one if needed?

If this is needed, I'd be partial to `with_filter(mut self, filter: Fn(Duration) -> bool)` so the caller can decide how to sample individual measurements

Reply:
This makes sense, I changed the API to be `with_filter(...)`!